### ISSUE TITLE
Trait lifetime bounds seem to be insufficient

### ISSUE THREAD

**Issue by davidhalter** [rust-lang/rust#92520]
I have a lifetime issue where the `'global: 'a` lifetime bound does not seem to be honored by an HRTB. Adding a generic default parameter fixes this. The original code was something like this:

```rust
trait Trait<'global: 'a, 'a> {
    fn foo(&self) -> u8 {
        todo!()
    }
}

struct Result<'global>(&'global ());

impl<'global> Result<'global> {
    fn run(
        &self,
        _callback: impl for <'a> FnMut(&dyn Trait<'global, 'a>)
    ) {
        todo!()
    }
}

fn main() {
    let var = ();
    let r = Result(&var);
    let mut x = 0;
    r.run(|some_trait| {
        // Both uncommenting this line and removing the 'global: 'a make this compile. 
        x += some_trait.foo();
    });
}
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e804b85d0603f9d514bdbb0300177b12)

I expected to see this compile.

Instead, lifetime errors are raised.

However if I change the trait signatures to `trait Trait<'global: 'a, 'a, HackyProof = &'a &'global ()>`, type inference ensures that the trait has the lifetime bounds `'db: 'a`. ([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a571480ec24b1712f6665e7d50e39def))

### Meta

Same issue for stable/nightly.

`rustc --version --verbose`:
```
rustc 1.59.0-nightly (c14569225 2022-01-01)
```

**Comment by compiler-errors** [comment]
Just for the record, this change also makes it compile (that is, explicitly naming the lifetime):

```diff
impl<'global> Result<'global> {
    fn run(
        &self,
-        _callback: impl for <'a> FnMut(&dyn Trait<'global, 'a>)
+        _callback: impl for <'a> FnMut(&'a dyn Trait<'global, 'a>)
    ) {
        todo!()
    }
}
```

Desugaring `for<'a> FnMut(&dyn Trait<'global, 'a>` into `for<'a, 'b> FnMut(&'b Trait<'global, 'a>)` seems to yield the same error.

Interestingly, though possibly unrelated, changing that `foo` call to take `&'a self` also leads to some interesting (and quite unhelpful) errors: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f285eafb6e9ed34ef47fdcc0b6f1b59b)

**Comment by compiler-errors** [comment]
Also interesting, if we put the reference into a wrapper type, it compiles as well: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=62858ebb5f1aceedf1ba655e5cf71865
